src/Swarm/Game/Scenario/Topography/Navigation/Portal.hs

```diff
@@ -97,24 +153,109 @@ validateNavigation upperLeft unmergedWaypoints portalDefs = do
+  MonadFail m =>
+  Navigation (M.Map SubworldName) WaypointName ->
+  m (M.Map (Cosmo Location) (AnnotatedDestination Location))
+validatePortals (Navigation wpUniverse partialPortals) = do
+  portalPairs <- forM (M.toList partialPortals) $ \(portalEntrance, AnnotatedDestination isConsistent portalExit@(Cosmo swName (WaypointName rawExitName))) -> do
+    firstExitLoc :| otherExits <- getLocs portalExit
+    unless (null otherExits)
+      . fail
+      . T.unpack
+      $ T.unwords
+        [ "Ambiguous exit waypoints named"
+        , quote rawExitName
+        , "for portal"
+        ]
+    return (portalEntrance, AnnotatedDestination isConsistent $ Cosmo swName firstExitLoc)
+
+  ensureSpatialConsistency portalPairs
+
+  return $ M.fromList portalPairs
+ where
+  getLocs (Cosmo swName wpWrapper@(WaypointName exitName)) = do
+    subworldWaypoints <- case M.lookup swName wpUniverse of
+      Just x -> return x
+      Nothing ->
+        fail $
+          T.unpack $
+            T.unwords
+              [ "Could not lookup waypoint"
+              , quote exitName
+              , "for portal exit because subworld"
+              , quote $ renderWorldName swName
+              , "does not exist"
+              ]
+
+    failWaypointLookup wpWrapper $
+      M.lookup wpWrapper subworldWaypoints
+
+-- | A portal can be marked as \"consistent\", meaning that it represents
+-- a conventional physical passage rather than a \"magical\" teleportation.
+--
+-- If there exists more than one \"consistent\" portal between the same
+-- two subworlds, then the portal locations must be spatially consistent
+-- between the two worlds. I.e. the space comprising the two subworlds
+-- forms a "conservative vector field".
+--
+-- Verifying this is simple:
+-- For all of the portals between Subworlds A and B:
+-- * The coordinates of all \"consistent\" portal locations in Subworld A
+--   are subtracted from the corresponding coordinates in Subworld B. It
+--   does not matter which are exits vs. entrances.
+-- * The resulting \"vector\" from every pair must be equal.
+ensureSpatialConsistency ::
+  MonadFail m =>
+  [(Cosmo Location, AnnotatedDestination Location)] ->
+  m ()
+ensureSpatialConsistency xs =
+  unless (null nonUniform) $
+    fail $
+      unwords
+        [ "Non-uniform portal distances:"
+        , show nonUniform
+        ]
+ where
+  consistentPairs :: [(Cosmo Location, Cosmo Location)]
+  consistentPairs = map (fmap cosmoLocation) $ filter (enforceConsistency . snd) xs
+
+  interWorldPairs = filter (uncurry ((/=) `on` view subworld)) consistentPairs
+  normalizedOrdering = map normalizePairOrder interWorldPairs
+
+  normalizePairOrder pair =
+    if uncurry ((>) `on` view subworld) pair
+      then Negative $ swap pair
+      else Positive pair
+
+  tuplify = both (view subworld) &&& both (view planar)
+
+  nest ::
+    Signed (b, a) ->
+    (b, Signed a)
+  nest = \case
+    Positive x -> fmap Positive x
+    Negative x -> fmap Negative x
```

Comment:
I thought `sequenceA` would work here instead, but the compiler doesn't like it.

Reply:
Resolved this question in b1a771e.